### PR TITLE
feat: improve error handling and parser

### PR DESCRIPTION
**Pull request by SofianD**

### CODE REVIEW

**Review by PierreDemailly — APPROVED** [review]

File: test/utils.spec.ts [hunk 1]
```diff
@@ -6,6 +6,17 @@ import stream from "stream";
...
+function toArrayBuffer(buffer: Buffer) {
+  const arrayBuffer = new ArrayBuffer(buffer.length);
+  const view = new Uint8Array(arrayBuffer);
+
+  for (let i = 0; i < buffer.length; ++i) {
+    view[i] = buffer[i];
+  }
+
+  return arrayBuffer;
+}
```
Comment (Dafyh) [inline test/utils.spec.ts:18]:
```ts
function toArrayBuffer(buffer: Buffer) {
  const { byteOffset, byteLength } = buffer;

  return buffer.buffer.slice(byteOffset, byteOffset + byteLength);
}
```

**Review by Dafyh — APPROVED** [review]

File: test/utils.spec.ts [hunk 2]
```diff
@@ -1,11 +1,16 @@
...
 import stream from "stream";
 
 // Import Internal Dependencies
 import * as Utils from "../src/utils";
 
+// function toArrayBuffer(buffer: Buffer) {
+//   const { byteOffset, byteLength } = buffer;
+
+//   return buffer.buffer.slice(byteOffset, byteOffset + byteLength);
+// }
```
Comment (Dafyh) [inline test/utils.spec.ts:12]: To remove ?

File: src/request.ts [hunk 3]
```diff
@@ -34,6 +35,14 @@ export interface RequestOptions {
   agent?: undici.Agent | undici.ProxyAgent | undici.MockAgent;
   // API limiter from a package like "p-ratelimit"
   limit?: InlineCallbackAction;
+  /**
+   * Default: `true`.
```
Comment (fraxken) [inline src/request.ts:39]:
```suggestion
   * @default: true
```

Comment (fraxken) [reply src/request.ts:39]: respecter la syntaxe JSDoc

File: src/class/undiciResponseHandler.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+   */
+  private async parseUndiciResponse<T>(): Promise<T | Buffer | string> {
+    const buffer = await this.getDecompressedBuffer();
+    const contentTypeHeader = this.response.headers["content-type"] as string;
+
+    if (!contentTypeHeader) {
+      return buffer;
+    }
+
+    let bodyAsString: string = void 0 as any;
```
Comment (Dafyh) [inline src/class/undiciResponseHandler.ts:121]:
```suggestion
    let bodyAsString = "";
```

File: src/common/errors.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+interface IGetErrorOptions<T> {
+  error?: Error;
+  message: keyof T;
+}
+
+// from myu-utils
+function taggedString(chaines: TemplateStringsArray, ...cles: string[] | number[]) {
+  return function cur(...valeurs: any[]): string {
+    const dict = valeurs[valeurs.length - 1] || {};
```
Comment (Dafyh) [inline src/common/errors.ts:11]:
```suggestion
    const dict = valeurs.at(-1)|| {};
```

File: src/common/errors.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable max-len */
+
+interface IGetErrorOptions<T> {
+  error?: Error;
+  message: keyof T;
+}
+
+// from myu-utils
+function taggedString(chaines: TemplateStringsArray, ...cles: string[] | number[]) {
+  return function cur(...valeurs: any[]): string {
```
Comment (Dafyh) [inline src/common/errors.ts:10]:
```suggestion
  return function cur(...values: any[]): string {
```

File: src/common/errors.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable max-len */
+
+interface IGetErrorOptions<T> {
+  error?: Error;
+  message: keyof T;
+}
+
+// from myu-utils
+function taggedString(chaines: TemplateStringsArray, ...cles: string[] | number[]) {
```
Comment (Dafyh) [inline src/common/errors.ts:9]: cles is `keys` ?

Comment (Dafyh) [reply src/common/errors.ts:9]: chaines is chains ?

File: src/request.ts [hunk 8]
```diff
@@ -78,18 +84,29 @@ export async function request<T>(
...
-    data: void 0 as any
+    data
   };
 
-  const data = await Utils.parseUndiciResponse<T>(requestResponse);
-  RequestResponse.data = data;
-
-  if (statusCode >= 400) {
-    throw Utils.toError(RequestResponse);
+  const shouldThrowOnHttpError = options.throwOnHttpError || true;
```
Comment (Dafyh) [inline src/request.ts:107]: This variable will always be set to `true` ?

File: src/request.ts [hunk 9]
```diff
@@ -78,18 +84,29 @@ export async function request<T>(
...
+  let data;
+  if (options.mode === "parse" || !options.mode) {
+    data = await responseHandler.getData<T>("parse");
+  }
+  else if (options.mode === "decompress") {
+    data = await responseHandler.getData("decompress");
+  }
+  else {
+    data = await responseHandler.getData("raw");
+  }
```
Comment (Dafyh) [inline src/request.ts:98]:
```suggestion
  if (options.mode === "parse" || !options.mode) {
    data = await responseHandler.getData<T>("parse");
  }
  else {
    data = await responseHandler.getData(options.mode);
  }
```

File: test/undiciResponseHandler.spec.ts [hunk 10]
```diff
@@ -0,0 +1,298 @@
+/* eslint-disable max-len */
+
+// Import Node.js Dependencies
+import { randomBytes } from "crypto";
+
+// Import Third-party Dependencies
+import { brotliCompressSync, deflateSync, gzipSync } from "zlib";
```
Comment (fraxken) [inline test/undiciResponseHandler.spec.ts:7]: zlib is a Node dependency

File: test/undiciResponseHandler.spec.ts [hunk 11]
```diff
@@ -0,0 +1,298 @@
+/* eslint-disable max-len */
+
+// Import Node.js Dependencies
+import { randomBytes } from "crypto";
```
Comment (fraxken) [inline test/undiciResponseHandler.spec.ts:4]: add prefix `node:`

**Review by fraxken — COMMENTED** [review]
I think you need to update the README to includes new docs like errors no?

File: src/utils.ts [hunk 12]
```diff
@@ -83,6 +47,14 @@ export function createHeaders(options: Partial<Pick<RequestOptions, "headers" |
   return headers;
 }
 
+export function isHttpieError(error) {
```
Comment (fraxken) [inline src/utils.ts:50]:
```suggestion
export function isHttpieError(error: unknown): error is HttpieError {
```

File: src/utils.ts [hunk 13]
```diff
@@ -83,6 +47,14 @@ export function createHeaders(options: Partial<Pick<RequestOptions, "headers" |
...
+  return error instanceof HttpieError;
+}
+
+export function isHTTPError(error) {
```
Comment (fraxken) [inline src/utils.ts:54]:
```suggestion
export function isHTTPError(error: unknown): error is HttpieOnHttpError {
```

File: test/HttpieOnHttpError.spec.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+      statusMessage: "Not Found",
+      data: null,
+      headers: {}
+    };
+
+    const error = new HttpieOnHttpError(reqResponse);
+    expect(error.name).toStrictEqual("HttpieOnHttpError");
+    expect(error).toMatchObject(reqResponse);
+  });
+});
```
Comment (fraxken) [inline test/HttpieOnHttpError.spec.ts:17]: This test do not provide any value (not testing anything).

**Pull request closed without merging**